Source/WebKit/UIProcess/gtk/AcceleratedBackingStoreDMABuf.cpp

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright (C) 2023 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE COMPUTER, INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE COMPUTER, INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "AcceleratedBackingStoreDMABuf.h"
+
+#if USE(EGL)
+#include "AcceleratedBackingStoreDMABufMessages.h"
+#include "LayerTreeContext.h"
+#include "WebPageProxy.h"
+#include <WebCore/IntRect.h>
+#include <WebCore/PlatformDisplay.h>
+#include <epoxy/egl.h>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/unix/UnixFileDescriptor.h>
+
+namespace WebKit {
+
+bool AcceleratedBackingStoreDMABuf::checkRequirements()
+{
+    static bool available;
+    static std::once_flag onceFlag;
+    std::call_once(onceFlag, [&] {
+        const char* disableDMABuf = getenv("WEBKIT_DISABLE_DMABUF_RENDERER");
+        if (disableDMABuf && strcmp(disableDMABuf, "0")) {
+            available = false;
+            return;
+        }
+
+        const auto& eglExtensions = WebCore::PlatformDisplay::sharedDisplay().eglExtensions();
+        available = eglExtensions.KHR_image_base
+            && eglExtensions.KHR_surfaceless_context
+            && eglExtensions.EXT_image_dma_buf_import
+            && eglExtensions.MESA_image_dma_buf_export;
```

Comment:
So I see MESA and assume this won't be available on NVIDIA proprietary graphics or other proprietary graphics drivers. Will it be available to everyone using mesa? Any example configurations that might result in getting the fallback path?

Reply:
The `MESA` bit in the name indicates where the extension originated, but does not prevent other drivers to implement it:

```
% strings libEGL_nvidia.so.525.89.02 | grep -i dmabuf      
eglQueryDmaBufFormatsEXT
eglQueryDmaBufModifiersEXT
eglExportDMABUFImageMESA
eglExportDMABUFImageQueryMESA
%
```

Reply:
It also requires `EGL_MESA_platform_surfaceless`. Situation is:
- `EGL_MESA_platform_surfaceless`
  - Supported in mesa since 2016 (don't know the exact version, it could be 12.x)
  - Enabled by default since mesa 17.1
  - Compiled unconditionally since mesa 20.2
- `EGL_MESA_image_dma_buf_export`
  - Initial support was added in 2014 (don't know the exact version, it cold be 10.x)
  - No compile option, so it's always avaialable